packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx

```diff
@@ -0,0 +1,239 @@
+  const {
+    hasError,
+    isClearable,
+    isDisabled,
+    isReadOnly,
+    name,
+    onBlur,
+    onChange,
+    onInitialLoadError,
+    onInputChange,
+    onLoadError,
+  } = props;
+  const [loadingError, setLoadingError] = React.useState(false);
+  const { formatMessage } = useIntl();
+  const { dataLocale, projectLanguages } = useApplicationContext((context) => ({
+    dataLocale: context.dataLocale,
+    projectLanguages: context.project?.languages,
+  }));
+  const prefixSearchFields = React.useMemo(
+    () => [`masterData.current.name.${dataLocale}`, 'key'],
+    [dataLocale]
+  );
+
+  const loadResourceWithPrefix = useDebouncedPromiseCallback(
+    useLoadWithPrefixSearchFields({
+      query: FetchProductByPrefix,
+      prefixSearchFields,
+    }),
+    TIME_TO_WAIT
+  );
+
+  const convertProductToOption = React.useCallback(
+    (product: TFlatProduct): TProductOption => {
+      const formattedName = formatLocalizedString(product, {
+        key: 'name',
+        locale: dataLocale,
+        fallback: NO_VALUE_FALLBACK,
+        fallbackOrder: projectLanguages,
+      });
+      return {
+        key: product.key,
+        label: formattedName,
+        value: product.id,
+      };
+    },
+    [projectLanguages, dataLocale]
+  );
+
+  const loadOptions = React.useCallback(
+    (inputValue) => {
+      return (
+        loadResourceWithPrefix({ inputValue })
+          .then(
+            ({ data }: { data?: TPrefixProductGraphqlProducts }) =>
+              data?.products?.results
+                ?.map(flattenProduct)
+                .map(convertProductToOption) ?? []
+          )
+          // @ts-ignore
+          .catch((graphQlErrors) => {
+            if (onLoadError) {
+              setLoadingError(true);
+              onLoadError(graphQlErrors);
+            }
+          })
+      );
+    },
+    [loadResourceWithPrefix, convertProductToOption, onLoadError]
+  );
+
+  const handleInputChange = React.useCallback(
+    (inputValue) => {
+      setLoadingError(false);
+      if (onInputChange) {
+        onInputChange(inputValue);
+      }
+    },
+    [onInputChange, setLoadingError]
+  );
+
+  // @TODO: handle error for initial load of non-existing product
+  // example scenarios
+  // GIVEN Merchant Center Product Details
+  // WITH custom field "product"
+  // AND custom field product has value "product-a"
+
+  // WHEN "product-a" no longer exists
+  // THEN set an error
+
+  // WHEN "product-a" fails to load
+  // THEN set an error
+  const prefetchSelectedProductQuery = useQuery(PrefetchProduct, {
+    variables: { id: props.value },
+    context: { target: GRAPHQL_TARGETS.COMMERCETOOLS_PLATFORM },
+    skip: !props.value,
+  });
+
+  const { error: prefetchError } = prefetchSelectedProductQuery;
+  React.useEffect(() => {
+    setLoadingError(true);
+    if (prefetchError && onInitialLoadError) {
+      onInitialLoadError(prefetchError);
+    }
+  }, [prefetchError, onInitialLoadError, setLoadingError]);
```

Comment:
Be careful with this!

If the props change their reference on each re-render, you end up in an infinite loop here.

I bumped into a similar issue once: https://github.com/commercetools/merchant-center-application-kit/pull/1567

I suggest to exclude `onInitialLoadError` from the dependencies, as the other values are controlled by us.